[Wu Jingya] iP

wujy28 opened the pull request:
# Moira<sup>TM</sup>
> "Your mind is not your friend." - Matt Berninger, The National

Do you resonate with the phrase, "No thoughts, head empty", and constantly find yourself forgetting crucial things? Well, today's your lucky day, as Moira<sup>TM</sup> is here to help! Ask her to:
- Manage your ToDo list
- Tell you a joke (TBC)
- Degrade you **for fun** (TBC)

Simply follow the steps below and get acquainted with your buddy Moira<sup>TM</sup>!

1. Download the latest build [here](https://github.com/wujy28/ip/releases/tag/v0.1).
2. Open up your computer's command line interface (CLI).
3. Navigate to the folder containing the JAR application.
5. Run the command `java -jar duke.jar`.
6. Say hi to Moira! :hugs:

Try Moira<sup>TM</sup> out with these commands:

- [ ] `todo {task description}` to add a new task with no deadline
- [ ]  `deadline {task description} /by {date}` to add a new task with a deadline[^1]
- [ ]  `event {task description} /from {date} /to {date}` to add a new event with a start and end date and time[^1]
- [ ]  `list` to list out all tasks
- [ ]  `delete {index}` to delete a task numbered _index_ from the list
- [ ]  `mark {index}` to mark a task numbered _index_ on the list as complete
- [ ]  `unmark {index}` to mark a task numbered _index_ on the list as incomplete
- [ ]  `find {keyword}` to search for tasks containing the specified keyword
- [ ]  `bye` to exit the application

[^1]: Dates must be entered in the format yyyy-MM-dd HH:mm

Well, what are you waiting for?
Hit Moira<sup>TM</sup> up **RIGHT NOW** for the low, low price of **ZERO DOLLARS**!!
> "You can't get a better deal than this." - Moira<sup>TM</sup>, 2023

##
Here is the `main` method of the program 'cuz we all love Java, don't we!
```ruby
    public static void main(String[] args) {
        scanner = new Scanner(System.in);
        tasks = new TaskList();
        storage = new Storage(tasks, filePath);
        ui = new Ui(name);
        parser = new Parser(ui);
        run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

wujy28 committed:
Changed name of Chatbot and Added greet and exit methods

wujy28 committed:
Added echo functionality to ChatBot

wujy28 committed:
Added add-to-list functionality and listing command

wujy28 committed:
Added Task class and mark as done functionality

wujy28 committed:
Added tracking for Todo, Deadline and Event Tasks, as well as handled certain invalid user inputs

wujy28 committed:
Added testcases and expected behavior for automated UI Testing, fixed some typos

wujy28 committed:
Added delete functionality to the task list

wujy28 committed:
Added JavaDoc comments to all important classes

wujy28 committed:
Added TaskList class to manage task list

wujy28 committed:
Added save system to maintain task list between sessions

wujy28 committed:
Merge branch 'branch-Level-7'

wujy28 committed:
Format dates using LocalDateTime

wujy28 committed:
Merge branch 'branch-Level-8'

wujy28 committed:
Reorganize code into classes UI, Storage, Parser & TaskList for better OOP

wujy28 committed:
Reorganize classes into packages: tasks, parser, ui, storage

wujy28 committed:
Merge branch 'master' into add-gradle-support

wujy28 committed:
Merge branch 'add-gradle-support'

wujy28 committed:
Add certain classes for JUnit Testing

wujy28 committed:
Add testcase for parse method in Parser, fix bug in Parser where it crashes if empty space is parsed

wujy28 committed:
Add testcase for parse method in Parser to check valid bye command, Add methods in Duke class to facilitate testing

wujy28 committed:
Add testcases for loadTasksFromFile method in Storage class, add temporary equals methods in task classes to aid testing

wujy28 committed:
Merge branch 'branch-A-JUnit'

wujy28 committed:
Package the app as a JAR file

wujy28 committed:
Add JavaDoc comments to all existing classes

wujy28 committed:
Tweak code in all existing classes to comply with coding standards

wujy28 committed:
Add search by keyword functionality to chat bot

wujy28 committed:
Merge branch 'branch-A-JavaDoc'

wujy28 committed:
Resolve merge conflict by fixing code in Duke and Storage classes

wujy28 committed:
Resolve merge conflict by fixing code in Parser and Ui classes

wujy28 committed:
Fix bug where tasks variable name in TaskList method was not updated

yyyaohhh commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,265 @@
+import duke.tasks.ToDoTask;
+import duke.ui.Ui;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
```
It is recommended to import them in alphabetical order 😀

yyyaohhh commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,265 @@
+                    handleInvalidInput();
+                }
+            }
+        } else {
```
Perhaps you can try to avoid deep nesting, as mentioned in the Code Quality textbook, because this makes readers harder to keep track of the logic

yyyaohhh commented on src/main/java/duke/tasks/ToDoTask.java:
```diff
@@ -0,0 +1,56 @@
+package duke.tasks;
+
+import duke.tasks.Task;
```
There is no need to import the class if they are from the same package 😉

yyyaohhh commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,98 @@
+ */
+public abstract class Task {
+    private String description;
+    private boolean done;
```
According to the Java Coding Standard textbook, boolean values should be named to sound like booleans.
For example, in this case, `done` should be name to `isDone`.

yyyaohhh commented on src/test/java/duke/storage/StorageTest.java:
```diff
@@ -0,0 +1,61 @@
+package duke.storage;
```
Great job for dividing the classes into many specific packages! 👍

yyyaohhh reviewed (commented):
Good job for the hard work!
The code quality and coding standards are generally well done!
There are some small things to take note of, such as naming and importing.
Look forward to chatting with Moira! (*^_^*)

wujy28 committed:
Configure build.gradle to accommodate checkstyle

wujy28 committed:
Transfer GUI from tutorial to chatbot

wujy28 committed:
Implement complete GUI for chat bot

wujy28 committed:
Merge branch 'branch-Level-10-Modifications' into branch-Level-10

wujy28 committed:
Merge branch 'branch-Level-10'

wujy28 committed:
Add checkstyle to project

wujy28 committed:
Merge branch 'branch-A-Checkstyle'

wujy28 committed:
Stop tracking config directory

wujy28 committed:
Modify build information for JAR exports

JAR applications exported currently adopt a file name duke.jar.

The chatbot's name is not Duke anymore, but Moira.

Therefore, the name has been changed to moira.jar for consistency.

wujy28 committed:
Modify code to adhere to coding standards

The code has clear violations of coding standard.

This makes the code look unprofessional.

Now the code mostly adheres to the coding standards.

More edits will be made when making changes to improve code quality.

wujy28 committed:
Improve the code quality in all classes

The code has a lot of code quality issues, most notably deep nesting and missing default cases,
causing the code to be very hard to follow, unreadable and unsafe.

Refactoring the code in compliance with code quality guidelines will help make the code
more reader-friendly and accessible for other programmers.

Let's,
- Remove deep nesting in Parser class to make the happy path more noticeable
- Change boolean names to sound more like booleans (e.g. done to isDone)
- Simplify complex statements involving conditional operators
- Add the default case to the switch block in the Storage class

wujy28 committed:
Add Java assertions in Duke, Parser and Storage classes

The code runs under certain important assumptions that are currently unchecked.

Adding assertions at various points in the code can detect potential runtime errors
arising from mistakes that programmers have made in the code.

Let's make use of the Java assert feature to increase the safety of our code.

wujy28 committed:
Merge pull request #1 from wujy28/branch-A-CodeQuality

Merge branch-A-CodeQuality

wujy28 committed:
Update branch-A-Assertions with master branch for merging

wujy28 committed:
Merge pull request #2 from wujy28/branch-A-Assertions

Merge branch-A-Assertions

wujy28 committed:
Set up Continuous Integration using gradle.yml file

wujy28 committed:
Remove I/O Redirection Test segments in gradle.yml

The gradle.yml file contains 3 segments at the end for I/O Redirection Tests.

Since we are not using a text-based UI anymore, let's remove them.

wujy28 committed:
Add sort task list functionality to chat bot

The current chat bot manages a list with tasks in the order of
the date they were added.

Allowing users to sort their task list by a specified sort
order can help them better organize their tasks.

Let's,
- Add a command to sort the task list
- Allow sorting by description, deadline(date) or task type

Representing sort order using a nested enum in the
TaskList class allows for easy reference of
available sort order types from other classes and
maintains encapsulation of information related to sorting.

wujy28 committed:
Add and update JavaDoc comments

wujy28 committed:
Merge branch 'branch-C-sort'

wujy28 committed:
Update testing to accommodate to GUI

The existing ParserTest class tests a method in a way
that is only for text-based UI programs.

The chat bot now uses a GUI for accepting user input.

Let's remove this ParserTest class and test another
method instead.

wujy28 committed:
Track previously untracked files so that they are available on Github

wujy28 committed:
Untrack duke.txt data file

wujy28 committed:
Reorganize test input files

wujy28 committed:
Untrack data directory

wujy28 committed:
Fix minor typo to comply with coding standards

wujy28 committed:
Merge branch 'branch-UpdatedTesting'

wujy28 committed:
Improve look of GUI Main Window and Dialogue Box

wujy28 committed:
Change font size of Send button text

wujy28 committed:
Merge branch 'branch-A-BetterGui'

wujy28 committed:
Start README.md draft

wujy28 committed:
Create CNAME

wujy28 committed:
Delete CNAME

wujy28 committed:
Add Ui screenshot to docs folder

wujy28 committed:
Update README.md: add table of contents to UG

wujy28 committed:
Update README.md: fix issues in table of contents

wujy28 committed:
Update README.md: add command summary table

wujy28 committed:
Fix warnings caused by JavaFX version incompatibility

wujy28 committed:
Update README.md: final changes

a55080993 commented:
_ Hello my friends, can we get to know each other? 
_

wujy28 committed:
Update README.md

Change README from generic Duke README to be more project-specific.
